src/utils/analytics.utils.ts

```diff
@@ -0,0 +1,117 @@
+import { EntityManager } from 'typeorm';
+import {
+  addMonths,
+  addDays,
+  startOfYear,
+  subYears,
+  format,
+  addYears,
+} from 'date-fns';
+import { EventStatus, Frequency, Months } from 'src/common/constants';
+import { User } from 'src/github/entities/user.entity';
+import { FilterDto } from 'src/analytics/dto/filter.dto';
+
+export const buildQueryBuilder = (
+  entityManager: EntityManager,
+  userId: string,
+  filter: FilterDto,
+) => {
+  const queryBuilder = entityManager
+    .getRepository(User)
+    .createQueryBuilder('user')
+    .leftJoinAndSelect('user.repo', 'repo')
+    .leftJoinAndSelect('repo.pullRequests', 'pullRequest')
+    .where('user.githubUserId = :githubUserId', { githubUserId: userId });
+
+  if (filter?.repoName) {
+    queryBuilder.andWhere('repo.name = :name', { name: filter.repoName });
+  }
+
+  return queryBuilder;
```

Comment:
PullProofAI

Consider adding validation for the `filter.repoName` input to ensure it is a non-empty string before applying it to the query builder.

```diff
+   if (filter?.repoName && filter.repoName.trim() !== '') {
+     queryBuilder.andWhere('repo.name = :name', { name: filter.repoName });
+   }
```



<!-- This is an auto-generated comment by PullProofAI -->